spdata/discover.py

```diff
@@ -0,0 +1,91 @@
+"""Methods providing functionality for dataset discovery across the filesystem
+
+Copyright 2018 Spectre Team
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+    
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import os
+
+from typing import Callable, Dict, List, Optional, Union
+from functools import partial, lru_cache
+from hashlib import sha256
+from glob import glob
+
+from .utility import as_readable, UnknownIdError
+from .common import DATA_ROOT, Name, Path
+
+MAX_JAVASCRIPT_SAFE_INT = 2 ** 53 - 1
+
+@lru_cache(maxsize=1024)
+def name_to_id(entity_name: str) -> int:
+    """Get ID for a name
+    Computes SHA256 hash of name and converts it to int value.
+
+    Args:
+        entity_name: name of an object
+
+    Returns:
+        out: unique identifier
+    """
+    hex_hash = sha256(entity_name.encode()).hexdigest()
+    return int(hex_hash, 16) % MAX_JAVASCRIPT_SAFE_INT
+
+def id_to_name(element_id: int) -> Name:
+    """Resolve element name by its id
+    Args:
+        element_id: id of the element in names list
+
+    Returns:
+        out: name of the element under given id
+    """
+    matching = [_name for _, _name in enumerate(d['name'] for d in get_datasets())
+                if name_to_id(_name) == element_id]
+    if not matching:
+        raise UnknownIdError(element_id)
+    return matching[0]
+
+def dataset_path(dataset_name: Name) -> Path:
+    """Discover path to dataset
+    Args:
+        dataset_name: name of the dataset
+
+    Returns:
+        out: path to the dataset file
```

Comment:
There are type annotations already here and in several other docstrings; I would leave out `out` altogether, as in [here](http://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html):
```py
def function_with_pep484_type_annotations(param1: int, param2: str) -> bool:
    """Example function with PEP 484 type annotations.

    Args:
        param1: The first parameter.
        param2: The second parameter.

    Returns:
        The return value. True for success, False otherwise.

    """
```